### PR TITLE
[박수아] 5주차 과제 제출합니다.

### PR DESCRIPTION
**Pull request by sssuuuaaa**
CS와 관련된 개발 지식 또는 언어, 프레임워크 소개를 주제로 글을 작성했습니다.
블로그링크 : https://study-csa.tistory.com/7

### CODE REVIEW

File: sssuuuaaa/Week05/GCP로 ML 서버 연동 및 배포하기.md [hunk 1]
```diff
@@ -0,0 +1,114 @@
+# CS와 연관된 개발 경험 정리 글
+
+자체적인 스케줄링 알고리즘을 개발하여 최적의 항만 화물차 입항 스케줄링 시스템에 대한 프로젝트를 진행했었다. 이때 ML 서버에 연동하는 과정을 경험했는데 이 과정에 대해 서술하려고 한다.
```
Comment (Youthhing) [inline sssuuuaaa/Week05/GCP로 ML 서버 연동 및 배포하기.md:3]: 도입부에 글의 목적을 설명하는거 아주 좋습니다 👍 
아래 아키텍처 사진도 직관적이어서 보기 좋네요!

File: sssuuuaaa/Week05/GCP로 ML 서버 연동 및 배포하기.md [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+
+### 1. 모델 훈련 및 저장
+
+먼저 PyTorch를 사용하여 로컬 환경이나 GCP의 AI Platform Notebooks에서 모델을 훈련한다. 모델을 `.pt` 파일 형식으로 저장하여 배포 준비한다.
+
+```python
+
+# 모델 훈련 후 저장 예시
+torch.save(model.state_dict(), 'model.pth')
+```
```
Comment (Youthhing) [inline sssuuuaaa/Week05/GCP로 ML 서버 연동 및 배포하기.md:57]: 글에는 .pt 형식인데 실제 코드 파일명은 .pth 라고 되어있는데 둘이 같은 표현일까요?

File: sssuuuaaa/Week05/GCP로 ML 서버 연동 및 배포하기.md [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+
+service = googleapiclient.discovery.build('ml', 'v1')
+name = 'projects/{}/models/{}/versions/{}'.format(project_id, model_name, version_name)
+
+response = service.projects().predict(
+    name=name,
+    body={'instances': [your_input_data]}
+).execute()
+
+print(response['predictions'])
```
Comment (Youthhing) [inline sssuuuaaa/Week05/GCP로 ML 서버 연동 및 배포하기.md:112]: API를 호출하듯 올려놓은 학습된 모델이 있는 서버에 요청을 보내고 응답 받는 형식인가요?

**Review by Youthhing — APPROVED** [review]
AI 쪽은 기초 수업정도만 들은 잘 모르는 분야였지만 아키텍쳐 사진, 그림, 코드 주석등을 통해 잘 설명해주셔서 이해하기 편했습니다! 특히 도입부에 글의 목적을 설명해주시는게 너무 좋았어요:)

다만, 앞에서 설명해주신 항만시스템에 대한 내용보단 기초적인 틀을 바탕으로 GCP를 연결하는 과정을 설명하신 것 같은데 수아님이 적용하셨던 프로젝트에 대한 내용과 마지막에 사용하며 개인적으로 느낀점 2가지가 들어갔다면 더욱 수아님의 개발 경험이 정리된 글이 됐을 것 같아요! 

글쓰느라 고생하셨습니다!!

File: sssuuuaaa/Week05/GCP로 ML 서버 연동 및 배포하기.md (hunk 2 above)
Comment (sssuuuaaa) [reply sssuuuaaa/Week05/GCP로 ML 서버 연동 및 배포하기.md:57]: 네네! PyTorch의 확장자는 'pt'와 'pth'가 있는데 .pth의 경우 python path와 충돌 위험이 있어 보통은 .pt 확장자를 사용하는 것을 추천한다고 해요!! 못보고 넘어갔는데 저 코드도 model.pt로 바꿔두는게 더 좋겠네요..ㅎㅎ 감사합니다!

File: sssuuuaaa/Week05/GCP로 ML 서버 연동 및 배포하기.md (hunk 3 above)
Comment (sssuuuaaa) [reply sssuuuaaa/Week05/GCP로 ML 서버 연동 및 배포하기.md:112]: 네네 맞습니다!!

File: sssuuuaaa/Week05/GCP로 ML 서버 연동 및 배포하기.md [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+5. ML 워크플로 조정
+6. 아티팩트 구성
+7. 모델 모니터링
+
+![Untitled](CS%E1%84%8B%E1%85%AA%20%E1%84%8B%E1%85%A7%E1%86%AB%E1%84%80%E1%85%AA%E1%86%AB%E1%84%83%E1%85%AC%E1%86%AB%20%E1%84%80%E1%85%A2%E1%84%87%E1%85%A1%E1%86%AF%20%E1%84%80%E1%85%A7%E1%86%BC%E1%84%92%E1%85%A5%E1%86%B7%20%E1%84%8C%E1%85%A5%E1%86%BC%E1%84%85%E1%85%B5%20%E1%84%80%E1%85%B3%E1%86%AF%20f85ccd5f8b604719ba663098bdb7a28a/Untitled%204.png)
+
+**모델 준비 및 훈련**
+
+머신러닝 모델의 훈련은 GCP의 AI Platform Notebooks나 로컬 환경에서 진행할 수 있다. TensorFlow, PyTorch, Scikit-learn 등 다양한 프레임워크를 사용하여 모델을 훈련시킨 후, 모델을 SavedModel, .pt 파일, .joblib 파일 등의 형식으로 저장한다. 우리 프로젝트에서는 PyTorch를 사용했었다.
+
```
Comment (ZeroKgun) [inline sssuuuaaa/Week05/GCP로 ML 서버 연동 및 배포하기.md:42]: 모델의 훈련은 일정량 이후에 종료되는건가요?? 아니면 계속 수시로 진행하는건가요??

**Review by ZeroKgun — APPROVED** [review]
ML에 대한 경험은 4학년 때 들었었던 '기계학습' 과목이 전부였는데 Scikit-learn 등 짧은 경험이었어도 꺼져있던 기억들이 생각났습니다. ML 쪽은 거의 문외한이라 술술 읽히진 않았지만 사용 라이브러리나 단계를 자세히 적어주셔서 찾아볼 거리가 생기는 글이었습니다. 고생하셨습니다!

File: sssuuuaaa/Week05/GCP로 ML 서버 연동 및 배포하기.md (hunk 4 above)
Comment (sssuuuaaa) [reply sssuuuaaa/Week05/GCP로 ML 서버 연동 및 배포하기.md:42]: 일정량 이후 종료됩니다!! 보통 epoch을 통해 데이터셋을 몇번 반복할지 결정하는데, 이 과정을 여러번 반복할수록 성능은 좋아져서 팀원들끼리 분배해서 최대한 많이 돌려서 결과값을 공유했습니다! 그렇게 해서 loss function 값이 목표만큼 떨어지면 훈련을 종료해요!

**Comment by MinJaeSon** [comment]
ML쪽은 제가 정말 무지해서.. 한 번 천천히 읽고 두 번 또 읽었네요ㅋㅋ 낯설지만 PyTorch를 사용해서 모델을 훈련시키고, 이를 GCP에 배포하여 서버에 연동하는 과정을 개요와 함께 그 밑에 흐름대로 설명해주셔서 그나마 이해하기 쉬웠던 것 같아요! 
제가 잘 모르거나 제가 공부하는 것과 크게 관련이 없는 것 같아도 제가 모르는 뭔가를 대충이라도 접해볼 수 있다는게 저는 참 좋더라구요. 그런 의미에서 이번 글도 읽어보면서 적게나마 ML쪽 지식도 얻고 좋네요ㅎ 수고하셨습니다~!
